lighthouse-logger/index.js

```diff
@@ -118,33 +174,145 @@ class Log {
+   * @template T, R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      let msg;
+      if (typeof opts.msg === 'string') {
+        msg = opts.msg;
+      } else if (typeof opts.msg === 'function') {
+        msg = opts.msg.apply(this, args);
+      }
+      if (!msg) {
+        throw new Error('expected msg');
+      }
+
+      let id;
+      if (typeof opts.id === 'string') {
+        id = opts.id;
+      } else if (typeof opts.id === 'function') {
+        id = opts.id.apply(this, args);
+      } else {
+        id = `lh:${originalFn.name}`;
+      }
+      if (!id) {
+        throw new Error('expected id');
+      }
+
+      const status = {msg, id};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.apply(this, args);
+      } catch (err) {
+        err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/g, '');
+        throw err;
+      }
+
+      if (result && typeof result.then === 'function') {
+        return result.then((/** @type {any} */ value) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          return value;
+        }).catch((/** @type {any} */ err) => {
+          err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/, '');
+          throw err;
+        });
+      } else {
+        Log.timeEnd(status, timeEndLogLevel);
+        return result;
+      }
+    };
+    return fn;
+  }
+  /* eslint-enable no-invalid-this */
+
+  /**
+   * Decorates (like timeDecorate) specified methods of a class.
+   * If decorating instance methods, use the class's prototype.
+   * If decorating static methods, use the class directly.
+   * @template {Object|Function} Class
+   * @template {keyof Class} Prop
+   * @param {Class} klass
+   * @param {{[key in Prop]: TimeDecorateOpts<Class, ArgumentTypes<IsFunction<Class[key]>>>}} methods
+   */
+  static timeDecorateClass(klass, methods) {
+    for (const [method, opts] of Object.entries(methods)) {
+      if (!opts.id) {
+        const className = (typeof klass === 'function' ? klass : klass.constructor).name;
+        opts.id = `lh:${className}:${method}`;
+      }
+      /** @type {IsFunction<Class[typeof method]>} */
+      const original = klass[method];
+      klass[method] = Log.timeDecorate(original, opts);
+    }
+  }
+
+  /**
+   * @param {string} title
+   * @param {ArrayAtLeastOne} args
+   */
   static log(title, ...args) {
     Log.events.issueStatus(title, args);
-    return Log._logToStdErr(title, args);
+    Log._logToStdErr(title, args);
```

Comment:
does this really change anything or was it just for ts?

Reply:
yeah, changes nothing. The return value was void.